staging/src/k8s.io/client-go/discovery/cached/memcache_test.go

```diff
@@ -98,35 +109,265 @@ func TestClient(t *testing.T) {
 			GroupVersion: "astronomy/v8beta1",
 			APIResources: []metav1.APIResource{{
-				Name:         "stars",
-				SingularName: "star",
+				Name:         "dwarfplanets",
+				SingularName: "dwarfplanet",
 				Namespaced:   true,
-				Kind:         "Star",
-				ShortNames:   []string{"s"},
+				Kind:         "DwarfPlanet",
+				ShortNames:   []string{"dp"},
 			}},
 		},
+		err: nil,
 	}
 	fake.lock.Unlock()
-
+	// We don't retry permanent errors, so it should fail.
+	_, err = c.ServerResourcesForGroupVersion("astronomy/v8beta1")
+	if err == nil {
+		t.Errorf("Expected error, got nil")
+	}
 	c.Invalidate()
+
+	// After Invalidate, we should retry.
 	r, err = c.ServerResourcesForGroupVersion("astronomy/v8beta1")
 	if err != nil {
 		t.Errorf("Unexpected error: %v", err)
 	}
-	if e, a := fake.resourceMap["astronomy/v8beta1"], r; !reflect.DeepEqual(e, a) {
+	if e, a := fake.resourceMap["astronomy/v8beta1"].list, r; !reflect.DeepEqual(e, a) {
+		t.Errorf("Expected %#v, got %#v", e, a)
+	}
+}
+
+func TestPartialRetryableFailure(t *testing.T) {
+	fake := &fakeDiscovery{
+		groupList: &metav1.APIGroupList{
+			Groups: []metav1.APIGroup{{
+				Name: "astronomy",
+				Versions: []metav1.GroupVersionForDiscovery{{
+					GroupVersion: "astronomy/v8beta1",
+					Version:      "v8beta1",
+				}, {
+					GroupVersion: "astronomy2/v8beta1",
+					Version:      "v8beta1",
+				}},
+			}},
+		},
+		resourceMap: map[string]*resourceMapEntry{
+			"astronomy/v8beta1": {
+				err: &errorsutil.StatusError{
+					ErrStatus: metav1.Status{
+						Message: "Some retryable error",
+						Reason:  metav1.StatusReasonServiceUnavailable,
+					},
+				},
+			},
+			"astronomy2/v8beta1": {
+				list: &metav1.APIResourceList{
+					GroupVersion: "astronomy2/v8beta1",
+					APIResources: []metav1.APIResource{{
+						Name:         "dwarfplanets",
+						SingularName: "dwarfplanet",
+						Namespaced:   true,
+						Kind:         "DwarfPlanet",
+						ShortNames:   []string{"dp"},
+					}},
+				},
+			},
+		},
+	}
+
+	c := NewMemCacheClient(fake)
+	if c.Fresh() {
+		t.Errorf("Expected not fresh.")
+	}
+	r, err := c.ServerResourcesForGroupVersion("astronomy2/v8beta1")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+	if e, a := fake.resourceMap["astronomy2/v8beta1"].list, r; !reflect.DeepEqual(e, a) {
+		t.Errorf("Expected %#v, got %#v", e, a)
+	}
+	_, err = c.ServerResourcesForGroupVersion("astronomy/v8beta1")
+	if err == nil {
+		t.Errorf("Expected error, got nil")
+	}
+
+	fake.lock.Lock()
+	fake.resourceMap["astronomy/v8beta1"] = &resourceMapEntry{
+		list: &metav1.APIResourceList{
+			GroupVersion: "astronomy/v8beta1",
+			APIResources: []metav1.APIResource{{
+				Name:         "dwarfplanets",
+				SingularName: "dwarfplanet",
+				Namespaced:   true,
+				Kind:         "DwarfPlanet",
+				ShortNames:   []string{"dp"},
+			}},
+		},
+		err: nil,
+	}
+	fake.lock.Unlock()
+	// We should retry retryable error, so no error is expected.
```

Comment:
nit: to be more clear, can you say "We should retry retryable error even without Invalidate() being called,"